metadata-ingestion/src/datahub/ingestion/source/pulsar.py

```diff
@@ -0,0 +1,726 @@
+    tenants_scanned: int = 0
+    namespaces_scanned: int = 0
+    topics_scanned: int = 0
+    filtered: List[str] = field(default_factory=list)
+    soft_deleted_stale_entities: List[str] = field(default_factory=list)
+
+    def report_tenants_scanned(self, topic_str: str) -> None:
+        self.tenants_scanned += 1
+
+    def report_namespaces_scanned(self, topic_str: str) -> None:
+        self.namespaces_scanned += 1
+
+    def report_topic_scanned(self, topic: str) -> None:
+        self.topics_scanned += 1
+
+    def report_dropped(self, topic: str) -> None:
+        self.filtered.append(topic)
+
+    def report_stale_entity_soft_deleted(self, urn: str) -> None:
+        self.soft_deleted_stale_entities.append(urn)
+
+
+@dataclass
+class PulsarSource(StatefulIngestionSourceBase):
+    config: PulsarSourceConfig
+    report: PulsarSourceReport
+    platform: str = "pulsar"
+
+    def __init__(self, config: PulsarSourceConfig, ctx: PipelineContext):
+        super().__init__(config, ctx)
+        self.config = config
+        self.report = PulsarSourceReport()
+        self.base_url = self.config.web_service_url + "/admin/v2"
+        self.tenants = self.config.tenants
+
+        self.session = requests.Session()
+        self.session.verify = self.config.verify_ssl
+        self.session.headers.update(
+            {
+                "Content-Type": "application/json",
+            }
+        )
+
+        if self._is_oauth_authentication_configured():
+            # Get OpenId configuration from issuer, e.g. token_endpoint
+            oid_config_url = (
+                "%s/.well-known/openid-configuration" % self.config.issuer_url
+            )
+            oid_config_response = requests.get(
+                oid_config_url, verify=False, allow_redirects=False
+            )
+
+            if oid_config_response:
+                self.config.oid_config.update(oid_config_response.json())
+            else:
+                logger.error(
+                    "Unexpected response while getting discovery document using %s : %s"
+                    % (oid_config_url, oid_config_response)
+                )
+
+            if "token_endpoint" not in self.config.oid_config:
+                raise Exception(
+                    "The token_endpoint is not set, please verify the configured issuer_url or"
+                    " set oid_config.token_endpoint manually in the configuration file."
+                )
+
+        # Authentication configured
+        if (
+            self._is_token_authentication_configured()
+            or self._is_oauth_authentication_configured()
+        ):
+            # Update session header with Bearer token
+            self.session.headers.update(
+                {"Authorization": f"Bearer {self.get_access_token()}"}
+            )
+
+        if (
+            self.is_stateful_ingestion_configured()
+            and not self.config.platform_instance
+        ):
```

Comment:
Can you explain in more detail, this part is from the KafkaSource

Reply:
I meant this could be there before creating the session in `__init__`. This would be fail-fast as checking config would be faster compared to connecting to server